src/Altinn.App.Core/Internal/Process/ProcessEngine.cs

```diff
@@ -1,138 +1,289 @@
+
+            events.Add(await GenerateProcessChangeEvent(InstanceEventType.process_EndEvent.ToString(), instance, now, user));
 
-            // find next task
-            List<ProcessElement> possibleNextElements = (await _flowHydration.NextFollowAndFilterGateways(processChange.Instance, validStartElement));
-            string? nextValidElement = ProcessHelper.GetValidNextElementOrError(null, possibleNextElements.Select(e => e.Id).ToList(),out ProcessError? nextElementError);
-            if (nextElementError != null)
+            // add submit event (to support Altinn2 SBL)
+            events.Add(await GenerateProcessChangeEvent(InstanceEventType.Submited.ToString(), instance, now, user));
+        }
+        else if (_processReader.IsProcessTask(nextElement?.Id))
+        {
+            var task = nextElement as ProcessTask;
+            currentState.CurrentTask = new ProcessElementInfo
             {
-                processChange.ProcessMessages = new List<ProcessChangeInfo>();
-                processChange.ProcessMessages.Add(new ProcessChangeInfo() { Message = $"Unable to goto next element due to {nextElementError.Code}-{nextElementError.Text}", Type = "Conflict" });
-                processChange.FailedProcessChange = true;
-                return processChange;
-            }
+                Flow = currentState.CurrentTask?.Flow + 1,
+                ElementId = nextElement!.Id,
+                Name = nextElement!.Name,
+                Started = now,
+                AltinnTaskType = task?.ExtensionElements?.AltinnProperties?.TaskType,
+                Validated = null,
+            };
+
+            events.Add(await GenerateProcessChangeEvent(InstanceEventType.process_StartTask.ToString(), instance, now, user));
+        }
+
+        // current state points to the instance's process object. The following statement is unnecessary, but clarifies logic.
+        instance.Process = currentState;
+
+        return events;
+    }
 
-            processChange.ProcessFlowElements.Add(nextValidElement!);
+    private async Task<InstanceEvent> GenerateProcessChangeEvent(string eventType, Instance instance, DateTime now, ClaimsPrincipal user)
+    {
+        int? userId = user.GetUserIdAsInt();
+        InstanceEvent instanceEvent = new InstanceEvent
+        {
+            InstanceId = instance.Id,
+            InstanceOwnerPartyId = instance.InstanceOwner.PartyId,
+            EventType = eventType,
+            Created = now,
+            User = new PlatformUser
+            {
+                UserId = userId,
+                AuthenticationLevel = user.GetAuthenticationLevel(),
+                OrgId = user.GetOrg()
+            },
+            ProcessInfo = instance.Process,
+        };
 
-            return await _processChangeHandler.HandleStart(processChange);
+        if (string.IsNullOrEmpty(instanceEvent.User.OrgId) && userId != null)
+        {
+            UserProfile up = await _profileService.GetUserProfile((int)userId);
```

Comment:
Seems like a http call that should have been done outside of this method, ideally done once for the entire user request and just passed in as a dependency.